### ISSUE TITLE
注意力熵

### ISSUE THREAD

**Issue by charonooooooo** [joshyZhou/AST#17]
你好，请问论文中的注意力熵是如何计算的

**Comment by joshyZhou** [comment]
Hi,

Thank you for your interest in our work. The attention entropy is calculated based on Eq. (9) in our paper. For convenience, we restate it here:: 

$Enctropy_{Att}=-\frac{1}{H}\sum\limits_h\frac{1}{L}\sum\limits_{ij}^{}Att_{ij}^{h,l}\ast{log^{Att_{ij}^{h,l}}}, $

Please feel free to contact me if you need any further help.

Yours,

Zhou

**Comment by charonooooooo** [comment]
请问有具体的代码实现吗，非常感激

**Comment by joshyZhou** [comment]
We share the code used. Please enjoy it:
```
def attention_entropy(attention):
    attention = attention + 1e-8  # to avoid getting NaNs (log0 = nan)
    return (-attention * torch.log(attention)).sum(-1).mean()
```

**Comment by charonooooooo** [comment]
我有个疑问，attention不一定大于0，torch.log(attention)会返回NAN

**Comment by joshyZhou** [comment]
The activation functions, regardless of whether they are ReLU () or Softmax(), ensure that the score of attention is not less than zero.

**Comment by charonooooooo** [comment]
感谢！！

**Comment by joshyZhou** [comment]
This issue is closed since it has been resolved.
